### PR TITLE
[WB-8176] Use Artifacts when put into run config. Accept a string to represent an artifact in the run config

### PR DESCRIPTION
**Pull request by KyleGoyette**
Fixes [WB-8176](https://wandb.atlassian.net/browse/WB-8176)

Description
-----------
What does the PR do?

[Design doc](https://www.notion.so/wandbai/Unifying-Run-Inputs-In-the-Config-Artifacts-In-the-Run-Config-03de8aac836b48c587e13e40b1086648)

Testing
-------
How was this PR tested?

Checklist
-------
- Name PR "[WB-NNNN][WB-MMMM] Add support for..." similar to entries in CHANGELOG.md
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: wandb/sdk/wandb_run.py [hunk 1]
```diff
@@ -971,8 +973,8 @@ def _config_callback(
     def _config_artifact_callback(
         self, key: str, val: Union[str, Artifact]
     ) -> Union[Artifact, public.Artifact]:
-        if isinstance(val, string_types) and val.startswith("wandb-artifact://"):
-            artifact_string = val[len("wandb-artifact://") :]
+        if isinstance(val, str) and val.startswith("wandb-artifact"):
```
Comment (vanpelt) [inline wandb/sdk/wandb_run.py]: Should be really rare, but I imagine a user could set a config value to "wandb-artifact-foo" and the parser would blow up.

File: wandb/util.py [hunk 2]
```diff
@@ -1467,9 +1467,63 @@ def check_dict_contains_nested_artifact(d: dict, nested: bool = False) -> bool:
...
+        raise ValueError(f"Invalid artifact string: {v}")
+    if len(parts) > 3:
+        # for now can't fetch paths but this will be supported in the future
+        # when we allow passing typed media objects
+        entity, project, name_and_alias_or_version = parts[:2]
+        return f"{entity}/{project}/{name_and_alias_or_version}"
+
+
+def parse_artifact_digest_string(v: str) -> str:
+    if not v.startswith("wandb-artifact-wandb-artifact-digest://"):
```
Comment (vanpelt) [inline wandb/util.py]: Double `wandb-artifact`?

File: wandb/sdk/wandb_run.py (hunk 1 above)
Comment (vanpelt) [reply wandb/sdk/wandb_run.py]: I see now, would be good to also check `and "://" in val`

**Review by vanpelt — APPROVED** [review]

File: wandb/util.py (hunk 2 above)
Comment (KyleGoyette) [reply wandb/util.py]: Ah, this whole function should be removed

File: wandb/sdk/wandb_run.py (hunk 1 above)
Comment (KyleGoyette) [reply wandb/sdk/wandb_run.py]: Fixed

File: wandb/sdk/wandb_init.py [hunk 3]
```diff
@@ -625,6 +629,11 @@ def init(self) -> Union[Run, RunDisabled, None]:  # noqa: C901
 
         self._wl._global_run_stack.append(run)
         self.run = run
+
+        # put artifacts in run config
+        for k, v in self.init_artifact_config.items():
```
Comment (raubitsj) [inline wandb/sdk/wandb_init.py:634]: Is there a reason why this is done after the Run is created?  (im guessing there is.. i just havent been able to mentally follow the flow in this case)

Comment (KyleGoyette) [reply wandb/sdk/wandb_init.py:634]: Yeah, since putting artifacts in the config automatically calls `use_artifact` we need the run to be upserted for those calls to not fail

File: wandb/sdk/wandb_run.py [hunk 4]
```diff
@@ -351,7 +353,11 @@ def __init__(
         self._quiet = self._settings.quiet
 
         self._output_writer = None
-        self._used_artifact_slots: List[str] = []
+        self._upgraded_version_message = None
+        self._deleted_version_message = None
+        self._yanked_version_message = None
+        self._used_artifact_slots: Dict[str, str] = {}
+        # self._used_artifact_slots: List[str] = []
```
Comment (raubitsj) [inline wandb/sdk/wandb_run.py:360]: please remove comment.. looks like you have a dict now.

File: wandb/sdk/wandb_init.py (hunk 3 above)
Comment (raubitsj) [reply wandb/sdk/wandb_init.py:634]: Got it. thanks.

**Review by stephchen — COMMENTED** [review]
the logic in `use_artifact` is getting kinda convoluted and hard to follow but otherwise looks fine to me
